catalyst-gateway-crates/c509-certificate/src/c509_cert.rs

```diff
@@ -0,0 +1,282 @@
+///     certificate_serial_number: UnwrappedBiguint::U64Value(123456),
+///     issuer,
+///     validity_not_before: 1672531200,
+///     validity_not_after: 1767225600,
+///     subject,
+///     subject_public_key_algo: PubKeyAlgoRegistry::Ed25519,
+///     subject_public_key: pub_key,
+///     extensions,
+///     issuer_sig_algo: SignatureAlgoRegistry::Ed25519,
+/// };
+///
+/// let c509_cert = generate_unsigned_c509_cert(tbs_cert);
+/// assert!(!c509_cert.is_empty());
+/// ```
+pub(crate) fn generate_unsigned_c509_cert(tbs_cert: TbsCertificate) -> Vec<u8> {
+    let mut buffer: Vec<u8> = Vec::new();
+
+    // Create a new encoder, passing the buffer by mutable reference
+    let mut encoder: Encoder<&mut Vec<u8>> = Encoder::new(&mut buffer);
+
+    let _unused = encoder.array(C509_CERTIFICATE_ELEMENTS);
+    // Encode the certificate type
+    let _unused = encoder.u8(tbs_cert.certificate_type as u8);
+    // Encode the certificate serial number
+    tbs_cert.certificate_serial_number.encode(&mut encoder);
+    // Encode issuer
+    tbs_cert.issuer.encode(&mut encoder);
+    // Encode validity_not_before
+    tbs_cert.validity_not_before.encode(&mut encoder);
+    // Encode validity_not_after
+    tbs_cert.validity_not_after.encode(&mut encoder);
+    // Encode subject
+    tbs_cert.subject.encode(&mut encoder);
+    // Encode subject public key
+    encode_subject_public_key(
+        tbs_cert.subject_public_key_algo,
+        tbs_cert.subject_public_key,
+        &mut encoder,
+    );
+    // Encode extensions
+    encode_extensions(tbs_cert.extensions, &mut encoder);
+    // Encode issuer signature algorithm
+    encode_signature_algorithm(tbs_cert.issuer_sig_algo, &mut encoder);
+    // issuerSignatureValue
+    encode_issuer_signature_value(vec![], &mut encoder);
+
+    buffer
+}
+
+// ---------------------------------------------------
+
+#[cfg(test)]
+mod test_c509_cert {
+
+    use super::*;
+    use crate::{
+        c509_enum::AttributesRegistry,
+        cbor_encoder::{AttributeRegistryValue, RelativeDistinguishedName, StringType},
+        extensions::{
+            alt_name::{GeneralName, GeneralNamesRegistry, GeneralNamesRegistryType},
+            Extension, ExtensionRegistry, ExtensionValue,
+        },
+    };
+
+    #[test]
+    fn test_generate_c509() {
+        let issuer: Name = vec![RelativeDistinguishedName {
+            name: AttributesRegistry::CommonName,
+            value: AttributeRegistryValue {
+                str_type: StringType::Utf8String,
+                str_value: "RFC test CA".to_string(),
+            },
+        }];
+        let subject: Name = vec![RelativeDistinguishedName {
+            name: AttributesRegistry::CommonName,
+            value: AttributeRegistryValue {
+                str_type: StringType::Utf8String,
+                str_value: "01-23-45-FF-FE-67-89-AB".to_string(),
+            },
+        }];
+        let pub_key_ed25519 = vec![
```

Comment:
better encode bytes using hex values and `hex` crate. Just for better readability